### PR TITLE
Adding components

### PR DESCRIPTION
**Pull request by anaverage-enri**
# Summary
Adding components that will be used throughout the project, including:
- AutoCompleteInput
- Button
- Header
- NumberInput
- ProductCard
- SearchInput

Utilities:
- Formatter.js
- Validator.js
- VariantsUtil.ts

And some more types for utilities:
- Brand.interface.ts
- OS.interface.ts
- index.ts

### CODE REVIEW

File: components/AutoCompleteInput.vue [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+      search: '',
+      selectedOption: {} as Option,
+      results: [] as Option[],
+      isOpen: false,
+      selectedIndex: 0,
+      isValid: true,
+    }
+  },
+  watch: {
+    selectedOption(oldOption: Option, newOption: Option): void {
```
Comment (azltdanh) [inline components/AutoCompleteInput.vue:62]: No need for `oldOption: Option, newOption: Option`

File: components/AutoCompleteInput.vue [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+<style scoped>
+.autocomplete {
+  display: flex;
+  align-items: center;
+  justify-content: space-between;
+  flex-wrap: wrap;
+}
+.autocomplete > label {
+  max-width: 30%;
+  min-width: 30%;
```
Comment (azltdanh) [inline components/AutoCompleteInput.vue:173]: Find another solution like `flex-basis`

File: components/AutoCompleteInput.vue [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+
+  /* width: 700px; */
+}
+.autocomplete > div {
+  max-width: 70%;
+  min-width: 70%;
+}
+.invalid {
+  border: 2px solid #e91e63 !important;
```
Comment (azltdanh) [inline components/AutoCompleteInput.vue:182]: refrain from using `!important`, try `autocomplete.invalid {}`

File: components/AutoCompleteInput.vue [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+}
+.autocomplete > div > input {
```
Comment (azltdanh) [inline components/AutoCompleteInput.vue:184]: prefer className instead of direct children.

File: components/Button.vue [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    },
+    label: {
+      type: String,
+      default: '',
+    },
+  },
+  computed: {
+    btnClass() {
+      return this.danger ? 'btn-danger' : 'btn-primary'
+    },
```
Comment (azltdanh) [inline components/Button.vue:32]: What if we have `btn-secondary`?

File: components/ProductCard.vue [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    </div>
+  </div>
+</template>
+
+<script lang="ts">
+import VariantsUtil from '~/services/VariantsUtil.js'
+import Formatter from '~/services/Formatter.js'
+import { Image } from '~/types/Image.interface'
+import { Product } from '~/types/Product.interface'
+import { Option } from '~/types/Option.interface'
```
Comment (azltdanh) [inline components/ProductCard.vue:28]: Create a `'~/types/index.ts'` to combine all the types

File: services/VariantsUtil.js [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+  getModelByKey(key) {
+    let model = {}
+    for (let x = 0; x < variants.os.length; x++) {
+      for (let y = 0; y < variants.os[x].brands.length; y++) {
+        for (let z = 0; z < variants.os[x].brands[y].models.length; z++) {
+          if (variants.os[x].brands[y].models[z].id === key) {
+            model = variants.os[x].brands[y].models[z]
+          }
+        }
+      }
```
Comment (azltdanh) [inline services/VariantsUtil.js:80]: Use `Array.filter` `Array.find` instead.

Comment (azltdanh) [reply services/VariantsUtil.js:80]: Also apply for other utils.

File: pages/index.vue [hunk 8]
```diff
@@ -1,20 +1,64 @@
...
+    },
+    brands() {
+      return Constants.BRANDS
+    },
+    models() {
+      return Constants.MODELS
+    },
+    colors() {
+      return Constants.COLORS
+    },
```
Comment (azltdanh) [inline pages/index.vue:56]: These items should not be in `'~/constants/MockupData'`

File: pages/index.vue [hunk 9]
```diff
@@ -1,20 +1,64 @@
...
 import Vue from 'vue'
+import { Constants } from '~/constants/MockupData'
 
-export default Vue.extend({})
+export default Vue.extend({
+  data() {
+    return {
+      search: '' as string,
+      number1: Number,
+      number2: Number,
```
Comment (azltdanh) [inline pages/index.vue:33]: Mixed with types vs values

File: components/SearchInput.vue [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+  width: 200px;
+  display: flex;
+  justify-content: space-between;
+  align-items: center;
+  padding: 0 20px;
+}
+.search input[type='text'] {
+  border: none;
+  width: 155px;
+  font-size: 14.9px;
```
Comment (azltdanh) [inline components/SearchInput.vue:34]: Why `14.9px`?

File: components/ProductCard.vue [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+    },
+    model(): string {
+      return (VariantsUtil.getModelByKey(this.product.model) as Option).value
+    },
+    color(): string {
+      return (VariantsUtil.getColorByKey(this.product.color) as Option).value
+    },
+    memory(): string {
+      return (VariantsUtil.getMemoryByKey(this.product.memory) as Option).value
+    },
```
Comment (azltdanh) [inline components/ProductCard.vue:63]: Exception handler?

File: components/ProductCard.vue [hunk 12]
```diff
@@ -0,0 +1,101 @@
+<template>
+  <div class="product-card">
+    <img
+      src="~/assets/img.png"
```
Comment (azltdanh) [inline components/ProductCard.vue:4]: Should be `src="image.src"`

File: components/NumberInput.vue [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+  },
+  computed: {
+    isDecimal() {
+      // only allow user to type dot when isDecimal prop is true
+      return this.decimal ? '' : 'return event.charCode >= 48'
+    },
+  },
+  created() {
+    this.$emit('input', this.initValue)
+  },
```
Comment (azltdanh) [inline components/NumberInput.vue:44]: Why we need this `$emit`?

File: components/Button.vue [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+  place-items: center;
+  transform: translate(var(--translate), var(--translate));
+  transition: transform 0.2s ease-in-out;
+}
+
+.btn > span:nth-child(1) {
+  opacity: 0.6;
+}
+
+.btn:hover > span:nth-child(2) {
```
Comment (azltdanh) [inline components/Button.vue:102]: Use exact className instead.

File: services/VariantsUtil.js [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+            models.push(model)
+          })
+        }
+      })
+    })
+    return models
+  },
+  getModelByKey(key) {
+    let model = {}
+    for (let x = 0; x < variants.os.length; x++) {
```
Comment (nguyentienphuc0505) [inline services/VariantsUtil.js:73]: Gộp 2 cái loop này forEach nhỉ

File: services/VariantsUtil.js [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+          }
+        }
+      }
+    }
+    return model
+  },
+  getOSVersionByKey(key) {
+    let os = {}
+    for (let i = 0; i < variants.os.length; i++) {
+      for (let j = 0; j < variants.os[i].versions.length; j++) {
```
Comment (nguyentienphuc0505) [inline services/VariantsUtil.js:87]: The same here, forEach

File: services/VariantsUtil.js [hunk 17]
```diff
@@ -0,0 +1,95 @@
+import * as variants from '~/variants.json'
+
+export default {
```
Comment (nguyentienphuc0505) [inline services/VariantsUtil.js:3]: define type return cho mấy cái methods này nhé em

File: components/AutoCompleteInput.vue [hunk 18]
```diff
@@ -0,0 +1,222 @@
...
+    items: {
+      type: Array as () => Array<Option>,
+      default: () => [],
+    },
+    value: {
+      type: Object as () => Option,
+      default: () => {},
+    },
+  },
+  data() {
```
Comment (nguyentienphuc0505) [inline components/AutoCompleteInput.vue:52]: {
    search: string, 
    selectedOption: Option,
    results: Option[],
    isOpen: boolean,
    selectedIndex: number,
    isValid: boolean
  }

File: components/AutoCompleteInput.vue [hunk 19]
```diff
@@ -0,0 +1,222 @@
...
+        // move to the bottom when the current highlighted result at top
+        this.selectedIndex = this.results.length - 1
+      }
+    },
+
+    // set the autocomplete result when user
+    // use up/down arrow key to select a suggested row
+    // and press Enter
+    onEnter(): void {
+      this.search = this.results[this.selectedIndex]?.value
```
Comment (nguyentienphuc0505) [inline components/AutoCompleteInput.vue:130]: this.results[this.selectedIndex]?.value || ''

File: components/Button.vue [hunk 20]
```diff
@@ -0,0 +1,115 @@
+<template>
+  <div class="btn-container">
+    <button class="btn" :class="type" @click="$emit('clicked')">
```
Comment (azltdanh) [inline components/Button.vue:3]: `@click="$emit('clicked')"` --> @click="$emit('click')"

File: components/AutoCompleteInput.vue [hunk 21]
```diff
@@ -0,0 +1,229 @@
...
+}
+.autocomplete-input {
+  position: relative;
+  font-size: 16px;
+  max-width: 100%;
+  min-width: 100%;
+  padding: 7px 25px;
+  border: 1px solid black;
+  border-radius: 10px;
+}
```
Comment (azltdanh) [inline components/AutoCompleteInput.vue:196]: Consider class order.

File: components/Button.vue [hunk 22]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+.btn span {
+  position: absolute;
+  top: 0;
+  left: 0;
+  height: 100%;
+  width: 100%;
+  clip-path: polygon(10% 0%, 100% 0%, 90% 100%, 0% 100%);
+}
```
Comment (azltdanh) [inline components/Button.vue:95]: Consider class order. Refrain from using `span` or html tag.

**Review by azltdanh — APPROVED** [review]
LGTM, still have some cmts

**Pull request merged**